perf: role instead of users in workflow actions

hasnain2808 opened the pull request:
Every time any workflow transition is possible we write one workflow action row per user of the role that is permitted to make the transition to the table
When any user performs the action we update his record and delete every other record from the table

A role is what decides if a user should be allowed to make a transition
If we create one row per role we are not losing any information
We are also updating who made the transition in the record

By making this change we will be reducing the number of records that are getting created which are then deleted too. 

- [x] use qb
- [x] ~~write a patch~~ it is not possible to predict what the older state was. Older state is needed to figure out the role to replace the user with. Handling it on the fly 
- [x] ~~Soft delete workflow and cleanup at once so that we do not keep on messing up the indexes all the time~~ Probably not needed anymore
- [x] Check if all the indexes that are set make sense
- [x] Add a child table to store all roles that can perform the transition
- [x] Find better name for Workflow Action Role 

How to test:
1. Create a complicated workflow with multiple roles and transitions
2. Check if the transitions are visible properly in the list view
3. Make transitions by logging in via different roles
4. The states should transition as expected

**Testing backward compatibility**
1. Find a site that has pending workflow transitions
2. Migrate the site to this branch
3. The older transitions should be visible in the workflow transition table 
4. It should be possible to complete those transitions

hasnain2808 committed:
perf: role in workflow actions

hasnain2808 committed:
refactor: use frappe.qb for update query

hasnain2808 committed:
fix: verify roles

hasnain2808 committed:
fix: handle old_style workflow actions

hasnain2808 committed:
chore: linting

hasnain2808 committed:
fix: handle first state

hasnain2808 committed:
chore: multiple statements on one line (colon)

hasnain2808 committed:
fix: delete stale workflow actions

hasnain2808 committed:
perf: soft delete workflow actions

hasnain2808 committed:
fix: patch to update default value

hasnain2808 committed:
fix: minor

ankush commented on frappe/workflow/doctype/workflow_action/workflow_action.py:
```diff
@@ -132,20 +138,95 @@ def return_link_expired_page(doc, doc_workflow_state):
+			WorkflowAction.status, 'Completed',
+		).set(
+			WorkflowAction.completed_by, user
+		).where(
+			WorkflowAction.reference_doctype == doc.get('doctype')
+		).where(
+			WorkflowAction.reference_name == doc.get('name')
+		).where(
+			WorkflowAction.role.isin(list(allowed_roles))
+		).where(
+			WorkflowAction.status == 'Open',
+		).where(
+			WorkflowAction.is_deleted == 0,
```
`&` allows `AND`ing multiple criterions 

`.where((a == b) & (c != d)))`

hasnain2808 replied:
chaining via where looks cleaner to me and more pythonic
No need for brackets to separate between & too
is what you are suggesting better?

ankush replied:
not necessarily better but it's closer to SQL query. SQL queries only have one `where`.

hasnain2808 committed:
refactor: better functions

fix: use int's instead of bool for queries
refactor: use frappe.qb for update queries

hasnain2808 committed:
fix: postgres support for limit

hasnain2808 commented:
Postgres does not allow setting limit without using order by
After adding order_by frappe.qb.update is giving issues

Aradhya-Tripathi commented on frappe/workflow/doctype/workflow_action/workflow_action.py:
```diff
@@ -132,20 +138,111 @@ def return_link_expired_page(doc, doc_workflow_state):
+		WorkflowAction.role.isin(list(allowed_roles))
+	).where(
+		WorkflowAction.status == 'Open',
+	).where(
+		WorkflowAction.is_deleted == 0,
+	).orderby('role').limit(1).run()
```
I don't think you can use ```orderyby``` in update, subquery or rowcount might be a better option.

hasnain2808 replied:
separated the select from update

hasnain2808 committed:
fix: remove limit out of update

hasnain2808 committed:
perf: better indexing

hasnain2808 committed:
refactor: use walrus for assignment

hasnain2808 committed:
test: test to approve using user

backwards compatibility test

hasnain2808 committed:
test: mariadb raises implicit commit exception

hasnain2808 committed:
fix: delete only users that are not needed

hasnain2808 committed:
test: postgres need quotes for fieldname

hasnain2808 committed:
ci: ignore semgrep for commit in test

gavindsouza commented on frappe/patches/v14_0/set_default_is_deleted_for_workflow.py:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2021, Frappe Technologies Pvt. Ltd. and Contributors
+# MIT License. See license.txt
+
+from __future__ import unicode_literals
```
```suggestion
# Copyright (c) 2022, Frappe Technologies Pvt. Ltd. and Contributors
# MIT License. See LICENSE

```

Is there a particular reason for adding this, or is this a part of some template? We're not supporting PY2 on v14 AFAIK so we can do without it.


---

btw, is this patch really required?

<img width="725" alt="Screenshot 2022-02-09 at 11 00 15 AM" src="https://user-images.githubusercontent.com/36654812/153127859-3ffaed6b-3a7b-4097-a665-d6b6b9fd18be.png">

If Frappe doesn't handle this correctly, there might have to be a fix in `DBTable.alter` instead.

gavindsouza commented on frappe/workflow/doctype/workflow/test_workflow.py:
```diff
@@ -93,6 +94,40 @@ def test_if_workflow_actions_were_processed(self):
 		frappe.set_user('Administrator')
 
+	def test_if_workflow_actions_were_processed_using_user(self):
+		frappe.db.delete("Workflow Action")
+		if not frappe.db.has_column('Workflow Action', 'user'):
+			# mariadb would raise this state would create an implicit commit
+			# if we do not commit before alter statement
+			# nosemgrep
+			frappe.db.commit()
+			frappe.db.multisql({
+				'mariadb': 'ALTER TABLE `tabWorkflow Action` ADD COLUMN user varchar(140)',
+				'postgres': 'ALTER TABLE "tabWorkflow Action" ADD COLUMN "user" varchar(140)'
+			})
```
PS: We could maintain these under the setUp method alone too, with an `if self._testMethodName == "test_if_workflow_actions_were_processed_using_user"`

hasnain2808 replied:
DOne

gavindsouza commented on frappe/workflow/doctype/workflow_action/workflow_action.py:
```diff
@@ -13,23 +13,30 @@
+
+	roles = frappe.get_roles(user)
+
+	if user == 'Administrator' or (doc.role in roles and doc.is_deleted == 0):
+		return True
+	else:
 		return False
```
```suggestion
	return user == "Administrator" or (not doc.is_deleted and doc.role in roles)
```

gavindsouza commented on frappe/workflow/doctype/workflow_action/workflow_action.py:
```diff
@@ -132,20 +138,110 @@ def return_link_expired_page(doc, doc_workflow_state):
+		WorkflowAction.user != user
+	).where(
+		WorkflowAction.status == "Open"
+	).run()
```
1. This and the following function seem to Falsy values if the user column exists or not. 
2. Should the user column be used at all? If the schema files don't define the Workflow Action DocType with a user table, why maintain the column at all? This column is one `bench trim-tables` away from being deleted. If it has information we need, transform and move it into the role column

hasnain2808 replied:
> 1. This and the following function seem to Falsy values if the user column exists or not.

This function would never be called if the user column does not exist. The if condition at the top is probably needed. I added it to make sure the function checks if it should run or not by itself. I think the if condition at the top should be removed.

> 2. Should the user column be used at all? If the schema files don't define the Workflow Action DocType with a user table, why maintain the column at all? This column is one `bench trim-tables` away from being deleted. If it has information we need, transform and move it into the role column

It is not possible to figure out which role gave the user the permission to perform an action just by looking at the user. We do not even store the previous state so no way to know how the document came to the state it is right now.

gavindsouza commented on frappe/workflow/doctype/workflow_action/workflow_action.py:
```diff
@@ -132,20 +138,110 @@ def return_link_expired_page(doc, doc_workflow_state):
+	).run()
+
+def update_completed_workflow_actions_using_user(doc, user=None):
+	user = user if user else frappe.session.user
```
```suggestion
	user = user or frappe.session.user
```

gavindsouza commented on frappe/workflow/doctype/workflow_action/workflow_action.py:
```diff
@@ -167,8 +263,10 @@ def get_next_possible_transitions(workflow_name, state, doc=None):
 	return transitions_to_return
 
 def get_users_next_action_data(transitions, doc):
+	roles = set()
```
Not sure how I feel about changing the return type from `Dict` to `Tuple[Dict, set]` but as long as this is an internal function, nobody should care.

gavindsouza reviewed (changes requested):
I'm not sure if we really need the `is_deleted` column. Are the benefits really worth adding an extra column and a daily job? And if so, why stop at this particular DocType and not add it everywhere? This feels like an anti-pattern.

surajshetty3416 commented:
Posting a discussion on this with @hasnain2808 

Let's create only one workflow action record with a list of roles that have access to perform that action. With this, we do not have to worry about deleting workflow actions and users will never see duplicate workflow actions if they have more than one role which has access to perform that workflow action.

hasnain2808 commented:
> I'm not sure if we really need the `is_deleted` column. Are the benefits really worth adding an extra column and a daily job? And if so, why stop at this particular DocType and not add it everywhere? This feels like an anti-pattern.

Although this might not be needed anymore for the sake of completion of the discussion I am adding the reasons here
1. The Workflow Action Table adds a record for each user of the role who can complete an action
2. In some cases there could be thousands of rows.
3. When any one of these users completes the action all the rows whose users do not perform the action are deleted so that they no longer show up in the active actions.
4. This can be a very expensive operation coz b+ trees reordering on disk and a lot of indexes are being recomputed. 
5. This was the reason for slowdowns being reported by users
5. Adding a flag does not do any of these actions. Inactive actions for all such workflows can be deleted at once via the job at a time when the system has less load

Soft deletes help when
1. There is a lot of data in the table
2. There is frequent deletions

hasnain2808 committed:
refactor: simpler conditions

Co-authored-by: gavin <gavin18d@gmail.com>

hasnain2808 committed:
https://github.com/frappe/frappe/pull/15789/

Co-authored-by: gavin <gavin18d@gmail.com>

hasnain2808 committed:
Merge branch 'develop' into save-role-in-workflow-action

hasnain2808 committed:
fix: delete not useful patch

hasnain2808 committed:
refactor: store permitted roles in a child table

hasnain2808 committed:
test: simplify tests

hasnain2808 committed:
test: clear-cache after DDl

hasnain2808 committed:
refactor: Workflow Action Role to Workflow Action Permitted Role

hasnain2808 committed:
fix: remove unused code

hasnain2808 committed:
Merge branch 'develop' into save-role-in-workflow-action

surajshetty3416 commented on frappe/workflow/doctype/workflow_action/workflow_action.py:
```diff
@@ -13,24 +13,44 @@
+	WorkflowAction = DocType("Workflow Action")
+	WorkflowActionPermittedRole = DocType("Workflow Action Permitted Role")
+
+	names_query = (frappe.qb.from_(WorkflowAction)
```
```suggestion
	permitted_workflow_actions = (frappe.qb.from_(WorkflowAction)
```

surajshetty3416 commented on frappe/workflow/doctype/workflow_action/workflow_action.py:
```diff
@@ -132,20 +151,83 @@ def return_link_expired_page(doc, doc_workflow_state):
-def update_completed_workflow_actions(doc, user=None):
+	allowed_roles = frappe.get_all('Workflow Transition',
+		fields='allowed',
+		filters=[['parent', '=', workflow],
+		['next_state', '=', workflow_state]],
```
```suggestion
		filters=[
			['parent', '=', workflow],
			['next_state', '=', workflow_state]
		],
```

surajshetty3416 commented on frappe/workflow/doctype/workflow_action/workflow_action.py:
```diff
@@ -132,20 +151,83 @@ def return_link_expired_page(doc, doc_workflow_state):
-		(user, doc.get('doctype'), doc.get('name'), user))
+	WorkflowAction = DocType("Workflow Action")
+
+	if not workflow_action or not workflow_action[0]:
+		return
```
```suggestion
	if not workflow_action:
		return
```
will there be a case where `workflow_action` will be `[[]]` 🤔

surajshetty3416 reviewed (changes requested):
@hasnain2808 after this change, no workflow action (old ones) will be visible to any user, right?

If we want to provide proper backward compatibility we might have to keep the `user` field (for some time) and show a workflow action list based on the user field as well.

hasnain2808 committed:
fix: remove not useful exit conditions

Co-authored-by: Suraj Shetty <13928957+surajshetty3416@users.noreply.github.com>

hasnain2808 committed:
chore: code formatting

Co-authored-by: Suraj Shetty <13928957+surajshetty3416@users.noreply.github.com>

hasnain2808 committed:
chore: better naming

Co-authored-by: Suraj Shetty <13928957+surajshetty3416@users.noreply.github.com>

hasnain2808 committed:
refactor: use f-strings

hasnain2808 committed:
fix: keep the user column intact

If this is not done no older workflow actions will be visible
This column will be removed after a while

hasnain2808 committed:
fix: show only open workflow actions
